src/connection/eagerlyConnect.ts

```diff
@@ -1,8 +1,23 @@
 import { Connector } from '@web3-react/types'
+import { useSyncExternalStore } from 'react'
 
 import { getConnection, gnosisSafeConnection, networkConnection } from './index'
 import { ConnectionType, selectedWalletKey, toConnectionType } from './types'
 
+let connectionReady: Promise<void> | true = true
+
+export function useConnectionReady() {
+  return useSyncExternalStore(
+    (onStoreChange) => {
+      if (connectionReady instanceof Promise) {
+        connectionReady.finally(onStoreChange)
+      }
+      return () => undefined
+    },
+    () => connectionReady === true
+  )
+}
+
```

Comment:
this is an interesting pattern. is this only needed because the connected state is updating outside the react lifecycle?

Reply:
yes, exactly that. `useSyncExternalStore` let's the react lifecycle subscribe to external updates and trigger renders accordingly.